Development/nmos/authorization_behaviour.cpp

```diff
@@ -0,0 +1,523 @@
+                            else
+                            {
+                                // client has not been registered to the Authorization server yet
+                                mode = client_registration;
+                            }
+                        }
+                        else
+                        {
+                            // scope-less client, not require to obtain access token
+                            mode = authorization_operation;
+                        }
+                    }
+                    else
+                    {
+                        // Should no further Authorization APIs be available or TTLs on advertised services expired, a re-query may be performed.
+                        mode = rediscovery;
+                    }
+                    break;
+
+                case client_registration:
+                    // register to the Authorization server to obtain client_id and client_secret (they can be found inside the client metadata)
+                    if (details::client_registration(model, authorization_state, load_ca_certificates, save_authorization_client, gate))
+                    {
+                        // client registered
+                        mode = with_read_lock(model.mutex, [&]
+                        {
+                            const auto& authorization_flow = nmos::experimental::fields::authorization_flow(model.settings);
+                            return (web::http::oauth2::experimental::grant_types::client_credentials.name == authorization_flow) ? authorization_operation : authorization_code_flow;
+                        });
+                    }
+                    else
+                    {
+                        // client registration failure, start authorization sequence again on next available Authorization server
+                        authorization_service_error = true;
+                        mode = request_authorization_server_metadata;
+                    }
+                    break;
+
+                case authorization_code_flow:
+                    if (details::authorization_code_flow(model, authorization_state, request_authorization_code, gate))
+                    {
+                        mode = authorization_operation;
+                    }
+                    else
+                    {
+                        // authorization code flow failure, start authorization sequence again on next available Authorization server
+                        authorization_service_error = true;
+                        mode = request_authorization_server_metadata;
+                    }
+                    break;
+
+                case authorization_operation:
+                    // fetch public keys
+                    // fetch access token in 1/2 token life time interval
+                    details::authorization_operation(model, authorization_state, load_ca_certificates, load_rsa_private_keys, false, gate);
```

Comment:
A little confused on the flow here.  Does `details::authorization_operation` block indefinitely on successful authentication?

Reply:
It continues to fetch public keys and access tokens in the defined intervals, until failed to obtain either keys or tokens.